File: apps/web/src/api/playground.ts
/**
 * 游乐园 API
 */
import request from '@/utils/request';

export type QuestionType =
  | 'audio'
  | 'meaning'
  | 'fill_blank'
  | 'context_cloze'
  | 'sentence_dictation';

export interface Question {
  id: string;
  type: QuestionType;
  word: string;
  meaning: string;
  hint: string;
  uk_phonetic?: string;
  us_phonetic?: string;
  sentence?: string;
  sentence_translation?: string;
  word_audio_url?: string;
  sentence_audio_url?: string;
}

export interface QuestionListResponse {
  questions: Question[];
  total: number;
}

export interface SubmitAnswerItem {
  question_id: string;
  question_type: QuestionType;
  word: string;
  meaning?: string;
  uk_phonetic?: string;
  us_phonetic?: string;
  sentence?: string;
  sentence_translation?: string;
  user_answer?: string;
  is_correct: boolean;
  attempts: number;
  showed_answer: boolean;
}

export interface SubmitPracticeRequest {
  answers: SubmitAnswerItem[];
  duration_seconds: number;
  max_streak: number;
}

export interface SubmitPracticeResponse {
  session_id: number;
  total: number;
  correct: number;
  wrong: number;
  skipped: number;
  accuracy: number;
  message: string;
  encouragement_zh: string;
  encouragement_en: string;
}

export interface PracticeSession {
  id: number;
  total_questions: number;
  correct_count: number;
  wrong_count: number;
  skipped_count: number;
  accuracy: number;
  duration_seconds: number;
  max_streak: number;
  created_at: string;
}

export interface PracticeHistoryResponse {
  sessions: PracticeSession[];
  total: number;
}

export interface PracticeStats {
  total_sessions: number;
  total_questions: number;
  total_correct: number;
  total_wrong: number;
  total_skipped: number;
  overall_accuracy: number;
  total_duration_minutes: number;
  best_streak: number;
}

export async function getPlaygroundQuestions(): Promise<QuestionListResponse> {
  return request.get('/playground/questions');
}

export async function submitPracticeResult(
  data: SubmitPracticeRequest,
): Promise<SubmitPracticeResponse> {
  return request.post('/playground/submit', data);
}

export async function getPracticeHistory(
  page = 1,
  pageSize = 10,
): Promise<PracticeHistoryResponse> {
  return request.get('/playground/history', {
    params: { page, page_size: pageSize },
  });
}

export async function getPracticeStats(): Promise<PracticeStats> {
  return request.get('/playground/stats');
}

File: apps/backend/app/schemas/playground.py
"""
游乐园训练相关数据模型。
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class QuestionResponse(BaseModel):
    id: str
    type: str
    word: str
    meaning: str
    hint: str
    uk_phonetic: str | None = None
    us_phonetic: str | None = None
    sentence: str | None = None
    sentence_translation: str | None = None
    word_audio_url: str | None = None
    sentence_audio_url: str | None = None


class QuestionListResponse(BaseModel):
    questions: list[QuestionResponse]
    total: int


class SubmitAnswerItem(BaseModel):
    question_id: str
    question_type: str
    word: str
    meaning: str | None = None
    uk_phonetic: str | None = None
    us_phonetic: str | None = None
    sentence: str | None = None
    sentence_translation: str | None = None
    user_answer: str | None = None
    is_correct: bool
    attempts: int = Field(default=1, ge=1, le=3)
    showed_answer: bool = False


class SubmitPracticeRequest(BaseModel):
    answers: list[SubmitAnswerItem]
    duration_seconds: int = Field(..., ge=0)
    max_streak: int = Field(default=0, ge=0)


class SubmitPracticeResponse(BaseModel):
    session_id: int
    total: int
    correct: int
    wrong: int
    skipped: int
    accuracy: float
    message: str
    encouragement_zh: str
    encouragement_en: str


class PracticeSessionResponse(BaseModel):
    id: int
    total_questions: int
    correct_count: int
    wrong_count: int
    skipped_count: int
    accuracy: float
    duration_seconds: int
    max_streak: int
    created_at: datetime


class PracticeHistoryResponse(BaseModel):
    sessions: list[PracticeSessionResponse]
    total: int


class PracticeStatsResponse(BaseModel):
    total_sessions: int
    total_questions: int
    total_correct: int
    total_wrong: int
    total_skipped: int
    overall_accuracy: float
    total_duration_minutes: int
    best_streak: int
